book-content/chapters/04-essential-types-and-annotations.md

```diff
@@ -0,0 +1,2366 @@
+
+### Exercises
+
+#### Exercise 1: Array Type
+
+Consider the following shopping cart code:
+
+```ts twoslash
+// @errors: 2353
+type ShoppingCart = {
+  userId: string;
+};
+
+const processCart = (cart: ShoppingCart) => {
+  // Do something with the cart in here
+};
+
+processCart({
+  userId: "user123",
+  items: ["item1", "item2", "item3"],
+});
+```
+
+We have a type alias for `ShoppingCart` that currently has a `userId` property of type `string`.
+
+The `processCart` function takes in a `cart` parameter of type `ShoppingCart`. Its implementation doesn't matter at this point.
+
+What does matter is that when we call `processCart`, we are passing in an object with a `userId` and an `items` property that is an array of strings.
+
+There is an error underneath `items` that reads:
+
+```
+Argument of type '{ userId: string; items: string[]; }' is not assignable to parameter of type 'ShoppingCart'.
+
+Object literal may only specify known properties, and 'items' does not exist in type 'ShoppingCart'.
+```
+
+As the error message points out, there is not currently a property called `items` on the `ShoppingCart` type.
+
+How would you fix this error?
+
+<Exercise title="Exercise 1: Array Type" filePath="/src/015-essential-types-and-annotations/028-arrays.problem.ts" resourceId="jUJqrXCHRph0Z4Fs6Ll3za"></Exercise>
+
+#### Exercise 2: Arrays of Objects
+
+Consider this `processRecipe` function which takes in a `Recipe` type:
+
+```ts twoslash
+// @errors: 2353
+type Recipe = {
+  title: string;
+  instructions: string;
+};
+
+const processRecipe = (recipe: Recipe) => {
+  // Do something with the recipe in here
+};
+
+processRecipe({
+  title: "Chocolate Chip Cookies",
+  ingredients: [
+    { name: "Flour", quantity: "2 cups" },
+    { name: "Sugar", quantity: "1 cup" },
+  ],
+  instructions: "...",
+});
+```
+
+The function is called with an object containing `title`, `instructions`, and `ingredients` properties, but there are currently errors because the `Recipe` type doesn't currently have an `ingredients` property:
+
+```
+Argument of type '{title: string; ingredients: { name: string; quantity: string; }[]; instructions: string; }' is not assignable to parameter of type 'Recipe'.
+
+Object literal may only specify known properties, and 'ingredients' does not exist in type 'Recipe'.
+```
+
+By combining what you've seen with typing object properties and working with arrays, how would you specify ingredients for the `Recipe` type?
+
+<Exercise title="Exercise 2: Arrays of Objects" filePath="/src/015-essential-types-and-annotations/029-arrays-of-objects.problem.ts" resourceId="YgFRxBViy44CfW0H2dToDx"></Exercise>
+
+#### Exercise 3: Tuples
+
+Here we have a `setRange` function that takes in an array of numbers:
+
+```ts twoslash
+// @errors: 2344
+import { Expect, Equal } from "@total-typescript/helpers";
+
+// @noUncheckedIndexedAccess: true
```

Comment:
You didnt include this flag up until now anywhere so it feels weird that this example is using it. Depending on how people will copy-paste this code and stuff they might not reach the same errors etc as u do with this flag